### ISSUE TITLE
Version 0.6 Development Status

### ISSUE THREAD

**Issue by pdeljanov** [pdeljanov/Symphonia#330]
Hey all,

As some may have noticed, most of the development activity on Symphonia is occurring in the `dev-0.6` branch. This is because I believe the 0.5.x series has reached a developmental dead-end where it is not possible to satisfactorily fix many of Symphonia's issues due to API constraints. Therefore, I've decided to focus my limited time on completing version 0.6 instead.

Version 0.6 includes many breaking changes to support popular requests such as making it easier to get audio out of Symphonia, or returning `None` when the end of the media is reached, but it also includes forward looking enhancements such as stubbing out video and subtitle decoding support.

### Development Plan

My development plan is as follows:

1. For the time being, I will be keeping 0.6 development in the `dev-0.6` branch until there are no more breaking changes being made.
2. Once the APIs are stabilised, I will cut a final 0.5.5 release from the mainline branch that includes some fixes for annoyances people have had. 
3. Once 0.5.5 is released, I will merge the 0.6 development branch into mainline and begin more thorough testing.
4. Once 0.6 has been tested (including some fuzzing time), it will be released.

### Timeline

I am hopeful for an early 2025 release, but some help from the community would improve the odds!

### Help Wanted

You can help the development and release of version 0.6 by picking up any of these tasks:

- [ ] Add fuzz testing harnesses to each format and codec crate.
- [ ] Begin playing around with version 0.6 and submitting feedback and suggestions on the new APIs. **Note: At this time, some of the new APIs are still in flux, but you can still get an idea, and different perspectives will be helpful to me.**
- [ ] Read, critique, fix, or update documentation.

I will add to this list over time...

Thanks all!

**Comment by crutchcorn** [comment]
Hey! I'm using the 0.6 branch against my [very very very small Tauri application](https://github.com/crutchcorn/moonsound), since it fixes many issues with FLAC currently

Just wanted to say that you're crushing it and I'm extremely happy as a downstream consumer of this package and that the API changes to 0.6 were pretty easy to adapt to

I'm still extremely green to Rust in general, so might not be a huge help, but if I end up with some spare cycles and find something critically blocking I'm happy to jump in

**Comment by caass** [comment]
One thing that could help with [no_std support](https://github.com/pdeljanov/Symphonia/issues/88) (which [seems](https://github.com/pdeljanov/Symphonia/pull/91) to have [stalled](https://github.com/pdeljanov/Symphonia/pull/244)) would be to move the core API away from `std::io` traits; for example, the tokio project has the [bytes](https://github.com/tokio-rs/bytes) crate which provides APIs similar to `ReadBytes`.

If you'd like to maintain the fallibility of the API (`bytes` panics implicitly instead), we could instead be generic over the error type, since [`Error`](https://doc.rust-lang.org/stable/core/error/trait.Error.html) has been stabilized in `core` for a couple releases now.

Unfortunately, the `Read` and `Write` (and `Seek`) traits are probably _not_ going to get moved to core [any time soon](https://github.com/rust-lang/rfcs/issues/2262), and I saw [here](https://github.com/pdeljanov/Symphonia/pull/244#issuecomment-1879827485) that it's impossible to copy those traits over due to licensing issues.

**Comment by sscobici** [comment]
Adding some points that would be good to address or consider before the 0.6 release:

1. **Packet's Track ID:** This is a unique identifier for the track. Previously, it was used as an index for the tracks array, but this is no longer the case, at least for MP4. Should we add a track_num field to the Packet structure?

2. **Packet PTS/DTS:** The introduction of Packet Presentation Time Stamp (PTS) and Decode Time Stamp (DTS) fields is needed.

3. **Packet Struct Design:** Should we keep the Packet struct for all types of packets (Audio, Video, Subtitles), or should we switch to an Enum, similar to how CodecParameters is structured?

4. **Track Metadata:** The Track struct currently lacks a TrackType field and a default flag. While analyzing CodecParams can be an alternative, it is not always convenient. Including a default flag would also provide useful information for users.

**Comment by hasezoey** [comment]
(symphonia 0.6 (516f1d4d88b944fb5975b83a2c9173fbc101b373), project `termusic`)
I could not find any changelog or migration guide, so i had to try and guess what the new values would be, mostly `symphonia-play` was helpful, but there are still some questions i have:
- What happened to [`ProbeResult`](https://docs.rs/symphonia-core/latest/symphonia_core/probe/struct.ProbeResult.html) and more specifically its `metadata` value? (which, IIRC, was part of the container's metadata, right?)
- What happened to [`SampleBuffer`](https://docs.rs/symphonia-core/latest/symphonia_core/audio/struct.SampleBuffer.html) and what should be used as the official alternative?
- Why is `AudioSpec` not `Copy` anymore?

Regarding `SampleBuffer`, i at first though i could use [`AudioBuffer`](https://github.com/pdeljanov/Symphonia/blob/516f1d4d88b944fb5975b83a2c9173fbc101b373/symphonia-core/src/audio/buf.rs#L68), but it is not actually possible because of `The trait "AsMut<[u8]>" is not implemented for "AudioBuffer<i16>", which is required by "&mut AudioBuffer<i16>: AsMut<[u8]>"` (for more see the stack traces below in the details), the only function that was not complaining about types was `decoded.copy_to(&mut buffer)`, but it errored at runtime with mismatching frame lengths.
But for now i could work-around it by using plain `Vec<i16>` similar to what `symphonia-play` uses and just cache the additional things like `frames()` from `decoded`.

<details>
<summary>Full compiler output</summary>

```txt
error[E0277]: the trait bound `AudioBuffer<i16>: AsMut<[u8]>` is not satisfied
   --> playback/src/rusty_backend/decoder/mod.rs:239:45
    |
239 |         decoded.copy_bytes_interleaved(&mut buffer);
    |                 ----------------------      ^^^^^^ the trait `AsMut<[u8]>` is not implemented for `AudioBuffer<i16>`, which is required by `&mut AudioBuffer<i16>: AsMut<[u8]>`
    |                 |
    |                 required by a bound introduced by this call
    |
    = note: required for `&mut AudioBuffer<i16>` to implement `AsMut<[u8]>`
note: required by a bound in `GenericAudioBufferRef::<'_>::copy_bytes_interleaved`
   --> /mnt/ssd/projects/rust/Symphonia/symphonia-core/src/audio/generic.rs:395:14
    |
392 |     pub fn copy_bytes_interleaved<Sout, Dst>(&self, dst: Dst)
    |            ---------------------- required by a bound in this associated function
...
395 |         Dst: AsMut<[u8]>,
    |              ^^^^^^^^^^^ required by this bound in `GenericAudioBufferRef::<'_>::copy_bytes_interleaved`
```
or
```txt
error[E0277]: the trait bound `AudioBuffer<i16>: AsMut<[_]>` is not satisfied
   --> playback/src/rusty_backend/decoder/mod.rs:239:48
    |
239 |         decoded.copy_to_slice_interleaved(&mut buffer);
    |                 -------------------------      ^^^^^^ the trait `AsMut<[_]>` is not implemented for `AudioBuffer<i16>`, which is required by `&mut AudioBuffer<i16>: AsMut<[_]>`
    |                 |
    |                 required by a bound introduced by this call
    |
    = note: required for `&mut AudioBuffer<i16>` to implement `AsMut<[_]>`
note: required by a bound in `GenericAudioBufferRef::<'_>::copy_to_slice_interleaved`
   --> /mnt/ssd/projects/rust/Symphonia/symphonia-core/src/audio/generic.rs:349:14
    |
346 |     pub fn copy_to_slice_interleaved<Sout, Dst>(&self, dst: Dst)
    |            ------------------------- required by a bound in this associated function
...
349 |         Dst: AsMut<[Sout]>,
    |              ^^^^^^^^^^^^^ required by this bound in `GenericAudioBufferRef::<'_>::copy_to_slice_interleaved`
```

where
```rs
let mut buffer = AudioBuffer::<i16>::new(decoded.spec().clone(), decoded.capacity());
```

---

</details>

Finally, there is also this warning when i use symphonia by path:
```txt
warning: field `bitrate` is never read
  --> /mnt/ssd/projects/rust/Symphonia/symphonia-bundle-mp3/src/common.rs:99:9
   |
96 | pub struct FrameHeader {
   |            ----------- field in this struct
...
99 |     pub bitrate: u32,
   |         ^^^^^^^
   |
   = note: `FrameHeader` has a derived impl for the trait `Debug`, but this is intentionally ignored during dead code analysis
   = note: `#[warn(dead_code)]` on by default

warning: `symphonia-bundle-mp3` (lib) generated 1 warning
```

Full result of my symphonia 0.6 try can be seen [in this compare](https://github.com/hasezoey/termusic/compare/8bdb62a8d27b1009e609191ad4c03c2b6111ba93..a6645428736f8b79b092ee7c89e0030d98409649) or [in this repository & branch](https://github.com/hasezoey/termusic/tree/symphonia06dev) (note that i might not keep that branch around forever).

**Comment by pdeljanov** [comment]
> One thing that could help with [no_std support](https://github.com/pdeljanov/Symphonia/issues/88) (which [seems](https://github.com/pdeljanov/Symphonia/pull/91) to have [stalled](https://github.com/pdeljanov/Symphonia/pull/244)) would be to move the core API away from `std::io` traits; for example, the tokio project has the [bytes](https://github.com/tokio-rs/bytes) crate which provides APIs similar to `ReadBytes`.
> 
> If you'd like to maintain the fallibility of the API (`bytes` panics implicitly instead), we could instead be generic over the error type, since [`Error`](https://doc.rust-lang.org/stable/core/error/trait.Error.html) has been stabilized in `core` for a couple releases now.
> 
> Unfortunately, the `Read` and `Write` (and `Seek`) traits are probably _not_ going to get moved to core [any time soon](https://github.com/rust-lang/rfcs/issues/2262), and I saw [here](https://github.com/pdeljanov/Symphonia/pull/244#issuecomment-1879827485) that it's impossible to copy those traits over due to licensing issues.

I think, for the sake of getting 0.6 this year, I will leave no_std to the next major release. :) 

Once we have 0.6 released, then we can start making internal changes (though not semver breaking), such as using `alloc`/`core` instead of `std`, transitioning lookup tables over to `phf`, etc. 

The `Error` trait looks like it stabilised fairly recently: 1.81. That's a bit new for Symphonia. I recently changed the MSRV to 1.77 and even that felt like pushing it. It'll probably be old enough for 0.7, though.

`ReadBytes` should cover all of `Read`'s use-cases, and I think we could also add our own `Seek` trait as well, or just use `MediaSource` for that.

**Comment by pdeljanov** [comment]
Updating `Packet` is one of the last changes that need to go in for 0.6. Provisioning any fields for video and subtitle tracks will be important.

> Adding some points that would be good to address or consider before the 0.6 release:
> 
> 1. **Packet's Track ID:** This is a unique identifier for the track. Previously, it was used as an index for the tracks array, but this is no longer the case, at least for MP4. Should we add a track_num field to the Packet structure?

The track ID should never be assumed to be a track index. It should always be matched against `Track::id`.

Perhaps it's better to replace it with the track index/number instead. What do you think? I've warmed up to having both in the packet, however, I'm not sure how useful the track ID itself is in practice, especially since it can essentially be a random number.

Additionally, I think we need to extend track IDs to `u64` since for some demuxer's like MKV it's being truncated right now.

> 2. **Packet PTS/DTS:** The introduction of Packet Presentation Time Stamp (PTS) and Decode Time Stamp (DTS) fields is needed.

Agreed. We can have both to replace `ts` right now. For audio we will simply have `pts` = `dts`.

> 3. **Packet Struct Design:** Should we keep the Packet struct for all types of packets (Audio, Video, Subtitles), or should we switch to an Enum, similar to how CodecParameters is structured?

Hard to say, but I'm not seeing a clear need right now. It depends what other fields we must carry for a video track.

> 4. **Track Metadata:** The Track struct currently lacks a TrackType field and a default flag. While analyzing CodecParams can be an alternative, it is not always convenient. Including a default flag would also provide useful information for users.

The `Track::flags` bitflags does have a default flag. Along with a few others. Agreed that `TrackType` would be easier.

**Comment by pdeljanov** [comment]
Hi @hasezoey,

Hopefully will have a migration guide and updated examples when the API changes are finalised. I do recommend running `cargo doc` and reading the updated API docs in the meantime, I think they are quite good.

> * What happened to [`ProbeResult`](https://docs.rs/symphonia-core/latest/symphonia_core/probe/struct.ProbeResult.html) and more specifically its `metadata` value? (which, IIRC, was part of the container's metadata, right?)

The metadata returned in the `ProbeResult` in 0.5.x was below was metadata the probe operation found _before_ the container was detected. This was typical for ID3v2 tags since it's just sort of glued on to the MPEG audio stream.

This metadata is now passed to the format reader and you can access it through the `metadata` function.

> * What happened to [`SampleBuffer`](https://docs.rs/symphonia-core/latest/symphonia_core/audio/struct.SampleBuffer.html) and what should be used as the official alternative?

You can now use the `copy_to`/`copy_bytes_to` family of functions on `AudioBuffer`, `AudioSlice`, `GenericAudioBuffer`, or `GenericAudioBufferRef` to extract the audio data in a format of your choice.

Here's a description of the different types of copy functions.

First, we have sample-oriented copy functions:

- `copy_to_slice_interleaved` copies all audio frames to a `&mut [S]` of the same or differing sample format in interleaved order.
- `copy_to_slice_planar` copies all audio planes to a  `&mut [&mut [S]]` of the same or differing sample format.
- `copy_to_vec_interleaved`/`copy_to_vec_planar` same as the other two, except takes a `&mut Vec<S>` /`&mut Vec<Vec<S>>` and resizes them accordingly. The slice versions need to be exactly sized, much like `copy_from_slice` in the standard library.

Next we have byte-oriented equivalents of the above:

- `copy_bytes_interleaved` copies interleaved audio to a `&mut [u8]` in the original sample format
- `copy_bytes_interleaved_as` same as above, except copies to a differing sample format.
- `copy_bytes_planar` copies all audio planes to a `&mut [&mut [u8]]` in the original sample format.
- `copy_bytes_planar_as` same as above, except copies to a differing sample format.
- `copy_bytes_to_vec_interleaved`/`copy_bytes_to_vec_interleaved_as`/`copy_bytes_to_vecs_planar`/`copy_bytes_to_vecs_planar_as` same as the other four except takes a `&mut Vec<u8>` /`&mut Vec<Vec<u8>>` and resizes them accordingly.

Therefore, you can simply replace (`Raw`)`SampleBuffer` with the equivalent code below. However, I recommend you read over the docs to see if something works better for you now. For example, if you have your own buffers, you could use the copy-to-slice functions.

```rust
// SampleBuffer
let mut samples: Vec<f32> = Default::default();
audio.copy_to_vec_interleaved(&mut samples);

// RawSampleBuffer
let mut raw_samples: Vec<u8> = Default::default();
audio.copy_bytes_to_vec_interleaved_as::<f32>(&mut raw_samples);
```

> * Why is `AudioSpec` not `Copy` anymore?

In previous versions, audio channels were represented using a set of bitflags. However, that was limiting for more advanced audio codecs, so channels can now also be represented as a `Vec` of channel labels. Since a `Vec` is not copyable, `AudioSpec` can no longer be `Copy`. However, if you don't use these more advanced channel layouts, then there is no additional overhead despite it being a "clone".

**Comment by hasezoey** [comment]
> This metadata is now passed to the format reader and you can access it through the metadata function.

Thanks for clarifying, this makes it a lot easier to deal with metadata. If i understand this correctly, you mean that it is included in this function [`Box<dyn FormatReader>::metadata`](https://github.com/pdeljanov/Symphonia/blob/93d16eb0b138f0502229cf34688edefcb2224236/symphonia-play/src/main.rs#L528).

> I do recommend running cargo doc and reading the updated API docs in the meantime, I think they are quite good.

That was what i was doing, though it only contained the *current state* not *how to migrate from old types*. (though yes, i have not run into any old docs yet)

> Therefore, you can simply replace (Raw)SampleBuffer with the equivalent code below. However, I recommend you read over the docs to see if something works better for you now. For example, if you have your own buffers, you could use the copy-to-slice functions.

That is basically what i am doing now, just wanted clarification if a plain `Vec<_>` is the replacement, or if i had missed anything.

For context, i had [made a wrapper buffer struct](https://github.com/hasezoey/termusic/blob/a6645428736f8b79b092ee7c89e0030d98409649/playback/src/rusty_backend/decoder/mod.rs#L331-L377), as we need more values than the raw samples (like `frame_len`).

TL;DR: the copy function looks like this:

```rs
/// Copy passed [`GenericAudioBufferRef`] into our Buffer
#[inline]
fn copy_buffers(&mut self, decoded: &GenericAudioBufferRef<'_>) {
    let required_capacity = decoded.byte_len_as::<i16>();
    self.buf.resize(required_capacity, 0);

    decoded.copy_to_vec_interleaved(&mut self.buf);
    self.frame_len = decoded.frames();
}
```

PS: completely unrealted, but because i dont know much about audio, would it be better to switch from `i16` to something higher like `f32` or any good material to read up on that?

**Comment by pdeljanov** [comment]
> > This metadata is now passed to the format reader and you can access it through the metadata function.
> 
> Thanks for clarifying, this makes it a lot easier to deal with metadata. If i understand this correctly, you mean that it is included in this function [`Box<dyn FormatReader>::metadata`](https://github.com/pdeljanov/Symphonia/blob/93d16eb0b138f0502229cf34688edefcb2224236/symphonia-play/src/main.rs#L528).

Yup.

> That is basically what i am doing now, just wanted clarification if a plain `Vec<_>` is the replacement, or if i had missed anything.

I see, yeah what you have looks like it'd work!

> PS: completely unrealted, but because i dont know much about audio, would it be better to switch from `i16` to something higher like `f32` or any good material to read up on that?

For all intents and purposes, `i16` is good enough since it has plenty of dynamic range. However, it's pretty typical for systems use `f32` internally for their audio processing pipelines and mixer so I'd recommend using that since many codecs decode to `f32` so you can avoid the `f32` -> `i16` -> `f32` conversion.

A classic intro video to digital media is this one: https://xiph.org/video/vid1.shtml.

If you want to get into more detail, I found this article to be quite detailed as well: https://mu.krj.st/wave/.

**Comment by edwloef** [comment]
I personally would appreciate being able to create iterators over the various `AudioBuffer`-style types, without having to put my own `Vec` in the middle and copying to and from that. I'm not entirely sure how that would look in planar form, but at least interleaved sounds like a fairly trivial implementation to me.

**Comment by GeoffClements** [comment]
Just been trying out 0.6 to see if there any major issues with my app. One major headache has cropped up; `mediasourcestream` now has a generic lifetime. I run the audio chain in a pulseaudio callback, so `mediasourcestream` is owned by the callback, only now the compiler can't tell if it lives long enough.

Not had much time to think of a solution yet but it seems like this one thing may cause a lot of revision.

**Comment by a1phyr** [comment]
@GeoffClements This changes allow strictly more possibilities than before. Everything that worked before should work the same with a `MediaSourceStream<'static>`. If you have lifetime issues, putting `'static` lifetimes everywhere the compiler tells you should fix them.

**Comment by GeoffClements** [comment]
@a1phyr  Ok, so my application creates an MSS, wraps it in a FormatReader and then moves the struct it into another thread for playing. When the track is played out then all the structs are torn down and the process repeats. Therefore the MSS does not have a static lifetime. I can try marking it as such but I doubt the compiler would be fooled.

When I add a generic lifetime as is now required by the type definition, moving to the playing thread is now impossible.

I can give it a try but i don't have time at the moment, so it may take a little time for me to get to it.

**Comment by GeoffClements** [comment]
Ok, I'm back in track. Finally got some time to do a bit of investigation plus a bit of reading. I was unaware as to the difference when `'static` is used as a reference lifetime and when used as a trait bound which contributed to my confusion.

Finally I have running code, but I have run in to a problem; `Track.codec_params` is constantly returning `None`, I've tried with flac, ogg/vorbis and mp3 files. I need to do more investigation first though. I'm pulling Symphonia from the latest dev-06 branch here on github so it's possible this area of code is still a work-in-progress.

**Comment by codedcosmos** [comment]
@pdeljanov Assuming no-one has started writing a migration guide.. I'm happy to try writing a migration guide at the same time as I update my application that heavily uses symphonia.

**Comment by sscobici** [comment]
0.6 is not ready yet. There were some proposals to improve public API.

PS.
I have some progress on video support. I introduced  PTS, DTS, adjusted mp4
demuxer, created
cbindings. Modified MPC-BE ( Windows media player for Windows) and
successfully played video stream.
There is still a lot of work to do. More demuxing features are required to
make it seekable I think.
But I'm not sure if symphonia needs video support right now. First of all
all audio issues needs to be resolved.

**Comment by baylesj** [comment]
> Hey all,
> 
> As some may have noticed, most of the development activity on Symphonia is occurring in the `dev-0.6` branch. This is because I believe the 0.5.x series has reached a developmental dead-end where it is not possible to satisfactorily fix many of Symphonia's issues due to API constraints. Therefore, I've decided to focus my limited time on completing version 0.6 instead.
> 
> Version 0.6 includes many breaking changes to support popular requests such as making it easier to get audio out of Symphonia, or returning `None` when the end of the media is reached, but it also includes forward looking enhancements such as stubbing out video and subtitle decoding support.
> 
> ### Development Plan
> My development plan is as follows:
> 
> 1. For the time being, I will be keeping 0.6 development in the `dev-0.6` branch until there are no more breaking changes being made.
> 2. Once the APIs are stabilised, I will cut a final 0.5.5 release from the mainline branch that includes some fixes for annoyances people have had.
> 3. Once 0.5.5 is released, I will merge the 0.6 development branch into mainline and begin more thorough testing.
> 4. Once 0.6 has been tested (including some fuzzing time), it will be released.
> 
> ### Timeline
> I am hopeful for an early 2025 release, but some help from the community would improve the odds!
> 
> ### Help Wanted
> You can help the development and release of version 0.6 by picking up any of these tasks:
> 
> * [ ]  Add fuzz testing harnesses to each format and codec crate.[ ]  Begin playing around with version 0.6 and submitting feedback and suggestions on the new APIs. **Note: At this time, some of the new APIs are still in flux, but you can still get an idea, and different perspectives will be helpful to me.**[ ]  Read, critique, fix, or update documentation.[ ]  Any help fixing decode or demuxing issues are always welcome.
> 
> I will add to this list over time...
> 
> Thanks all!

Have you considered a 0.5.6 release with the fuzzing and LICENSE issues included? Or is there a timeline for 0.6? Happy to help move things along however I can.

**Comment by pdeljanov** [comment]
I'm happy to do 0.5.x releases on request since there's only a trickle of bug fixes going into that branch these days. I do have some catch-up to do re: issues and PRs that have come in recently, but they look straightforward. So, I'll publish 0.5.6 once those have been resolved.

As for 0.6, my target is to publish a 0.6.0-alpha release by the end of the month. I don't usually do pre-release versions, but due to the breadth and depth of the changes in 0.6 I'm hoping the community can assist with testing and/or feedback regarding the API changes and such. We'll probably need 2 or more rounds of alpha/beta releases before a final release.

**Comment by aschey** [comment]
I'd be happy to help with any specific testing needed for 0.6. I've been using it pretty much every day for my audio player since 0.5 entered maintenance mode, so I've got plenty of experience with it already.

**Comment by baylesj** [comment]
Philip, that sounds great!

**Comment by pdeljanov** [comment]
Hello everyone,

I am happy to announce that I've published `v0.6.0-alpha.1` to crates.io. GitHub release notes TBD...

**Please note that this is an alpha release and there may be semver breaking changes between releases. You should not use this in production!** However, please do start experimenting with your migration over to 0.6 and report any issues.

Since v0.6.x is such a massive change from v0.5.x I decided *not* to merge `dev-0.6` back into `master`. Instead, I created a new `main` branch from `dev-0.6` and have made it the default branch. I then renamed the old `master` branch to `rel-0.5`. For those of you who have a cloned copy of Symphonia, you may need to adjust your local copy to correctly track the new branches.

Going forward, the `rel-*` branches will be sustaining branches. Therefore, if you'd like any fixes in a 0.5.x release, please target your PR towards `rel-0.5`.

Over the next few days I'll be closing PRs targeting v0.5 that contain major feature work. Likewise, I'll be closing issues that have been resolved in v0.6. Any open PRs that are targeting `dev-0.6` will be changed to target `main`. If you have any open PRs or Issues yourself and want to check if it's still required, that would be a big help!

As a heads up, I couldn't get to all the recently reported issues before publishing this release, however, they will be addressed in time. Likewise, some of the documentation may still be out-of-date, but I expect that all to be sorted out during this testing period.

Enjoy!

**Comment by edwloef** [comment]
I'm a bit surprised to see `all-meta` enabled by default, is that intended?

**Comment by pdeljanov** [comment]
> I'm a bit surprised to see `all-meta` enabled by default, is that intended?

Is that strange? It was always implicitly enabled up until now. However, since all the metadata formats are basically open standards, it makes sense for them to be enabled by default.

**Comment by M-Igashi** [comment]
I don't want you to raise your review cost, but I have tested the alpha1 in my app and found regression. Reported here: https://github.com/pdeljanov/Symphonia/issues/477

**Comment by hasezoey** [comment]
I have now done the symphonia 0.6.0 migration in our project (https://github.com/tramhao/termusic/pull/697).
It was a little annoying to do it without a proper guide / release notes and only being able to rely on the `symphonia-play` implementation, but it somehow worked out.

Though i am wondering why the ergonomics of a `From/Into (std)Duration` conversion (or at least helper functions) were not carried over for `Time`? (or am i missing something?)
Current workarounds i use:
```rust
// For seeking from a rodio::Source
//let pos: std::time::Duration;
let seek_to = units::Time::try_new(
    i64::try_from(pos.as_secs()).unwrap_or(i64::MAX),
    pos.subsec_nanos(),
)

// For total duration calc
let total_duration = track.num_frames.map(units::Duration::new).and_then(|dur| {
    track.time_base.map(|tb| {
        let ts = units::Timestamp::ZERO.saturating_add(dur);
        let (secs, nanos) = tb.calc_time_saturating(ts).parts();
        Duration::new(u64::try_from(secs).unwrap_or(0), nanos)
    })
})

// For elapsed duration
let elapsed = time_base
    .map(|tb| tb.calc_time_saturating(packet.pts()))
    .map(|time| {
        let (secs, nanos) = time.parts();
        Duration::new(u64::try_from(secs).unwrap_or(0), nanos)
    });
```

**Comment by pdeljanov** [comment]
Hello again,

I'm pleased to announce that the second alpha version, `v0.6.0-alpha.2`, has been published to crates.io.

My focus for this release has been driving down the number of open issues. From a high of 70 open issues it has now been reduced down to 42. My goal before the final release is to get to ~35 open issues, at which point most open issues would just be feature requests.

Aside from bug straight fixes, there has been a major rewrite of the internals of the MP4/M4A format reader. The WAVE reader also saw some more substantial changes. As such, please spend some time testing these two format readers.

My attention will now be focused more on the remaining open PRs, followed by some more bug fixes, and writing a migration guide. Depending on how that goes, I may make a beta release, or just publish a stable release.

---

@hasezoey Thanks for the feedback. I didn't add those because `std::time::Duration` is unsigned while Symphonia's `Duration` is signed so it would be fallible. I'll reconsider adding some helpers.

**Comment by M-Igashi** [comment]
Tested alpha.2 with [mp3rgain](https://github.com/M-Igashi/mp3rgain) (ReplayGain analyzer + lossless gain adjuster). Migrated from v0.5 → alpha.1 → alpha.2.

- MP3 analysis, AAC/M4A analysis, and decoding all work correctly
- Tested with 6 MP3 files (ID3v2 tags from 4KB to 8MB with embedded artwork) — no issues
- The sample clamping fix (#434) is confirmed working in alpha.2 — MP3 peak values above 1.0 are now correctly reported (was our main blocker on v0.5)
- All 52 unit/integration tests pass

No issues found. Thanks for the release!

**Comment by pdeljanov** [comment]
Symphonia v0.6 has been released! Thanks for your feedback!
